src/apply.jl

```diff
@@ -268,7 +268,7 @@ function optimise_p_q(
 
   fend = print_fidelity_loss ? fidelity(envs, p_cur, q_cur, p, q, o) : 0
 
-  if (print_fidelity_loss && real(fend - fstart) <= 0.0 && nsweeps >= 1)
+  if (print_fidelity_loss && real(fend - fstart) < -1e-15 && nfullupdatesweeps >= 1)
```

Comment:
You could use:
```julia
diff = real(fend - fstart)
if print_fidelity_loss && diff < -eps(diff) && nfullupdatesweeps >= 1
  # [...]
end
```
which would adjust the threshold for the element type instead of hardcoding `-1e-15`.